keps/sig-node/2371-cri-pod-container-stats/README.md

```diff
@@ -0,0 +1,762 @@
+    - [Feature Enablement and Rollback](#feature-enablement-and-rollback)
+    - [Rollout, Upgrade and Rollback Planning](#rollout-upgrade-and-rollback-planning)
+    - [Monitoring Requirements](#monitoring-requirements)
+    - [Dependencies](#dependencies)
+    - [Scalability](#scalability)
+    - [Troubleshooting](#troubleshooting)
+  - [Implementation History](#implementation-history)
+  - [Drawbacks](#drawbacks)
+  - [Alternatives](#alternatives)
+<!-- /toc -->
+
+# cAdvisor-less, CRI-full Container and Pod Stats
+
+## Release Signoff Checklist
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Kubelet relies on the [summary API](https://github.com/kubernetes/kubernetes/blob/release-1.20/staging/src/k8s.io/kubelet/pkg/apis/stats/v1alpha1/types.go) to provide node level stats, pod level stats and container level stats. 
+
+The [CRI API](https://github.com/kubernetes/cri-api) currently does not provide enough metrics to fully supply all the fields for the summary API.
+As such, today some summary API metrics are fetched from cAdvisor while others originate from calling into the CRI implementation.
+This results in an unclear origin of metrics, duplication of work done between cAdvisor and CRI, and performance implications. 
+
+This KEP aims to enhance the CRI API with enough metrics to be able to supplement the summary API directly from CRI alleviating the need for cAdvisor to collect pod and container level stats thereby ensuring only a single entity (CRI runtime) will collect container and pod level metrics.
+
+This KEP will eventually also aim to have the CRI implementation populate the container level stats needed for the `/metrics/cadvisor` endpoint (or deprecate that endpoint).
```

Comment:
I think having CRI populating all stats for kubelet summary is still reasonable, it is kubelet's responsibility today, and we can't deprecate it any time soon.

However, I prefer we don't let CRI provide all stats for `/metrics/cadvisor`, we should either deprecate it or keep it as it is.

Reply:
I don't think we can deprecate it. `/metrics/cadvisor` is basically part of the kube API at this point, and there are tons of observation infra that's built on top of it.

keeping it as-is would defeat part of the purpose of this KEP. we'd still hit the performance issues (two places would be scraping the data), and it would further confuse the source of the metrics (one endpoint comes from cAdvisor, one from CRI).

I don't know if the wording is clear, but the current idea is not to have the required metrics go through the CRI, but rather broadcasted directly from the CRI implementation (CRI impl->endpoint. rather than CRI impl->Kubelet->endpoint). This would allow for the metrics to continue to be supported, but avoid the hop

Reply:
> I don't think we can deprecate it. /metrics/cadvisor is basically part of the kube API at this point, and there are tons of observation infra that's built on top of it.

I think we can still deprecate it if:
1) We provide an alternative solution;
2) Give people long enough time to deprecate.

One possible solution is to:
1) Extract pod/container metrics from the current metrics exposed in `/metrics/cadvisor`, those do make sense for container runtimes to expose, and we clearly define the endpoint, the metrics required, the labels etc. in CRI, so that both containerd and cri-o can implement it;
2) Leave `/metrics/cadvisor` in kubelet, mark it as deprecated and provide option to disable it;
3) Provide a daemonset cadvisor solution in `kubernetes-sigs`, and point people to it in the deprecation notice.
4) Deprecate the `/metrics/cadvisor` endpoint eventually after >1? year.

If people just want pod/container metrics, they can use the metrics endpoint exposed by the container runtime, but may need to make some format change.

If people want other cadvisor metrics, e.g. systemd service metrics?, they should run the standalone cadvisor daemonset.

By doing this, we can move one step further with the metrics pipeline refactoring. Or else it will be even harder to deprecate if we ask people to implement the same metrics endpoint in the container runtime.

Reply:
I had initially been more gun-ho on supporting `/metrics/cadvisor` endpoint itself (as awkward as the name would be). However, since there are already breaking changing in this KEP for `/metrics/cadvisor` --mainly the URL they're exposed from-- we have an opportunity for us to make more breaking changes.

Thus, I've changed my mind and think not mandating the endpoint name is good, as long as we find a way to make sure the CRI impls support the required metrics. To me that is no different than changing the url that the endpoint would need to be implemented on. 

I think it would be ideal to have the label values unchanged from cadvisor, so a user could plug and play the old URL+endpoint with the new one. Do you see that working with the existing metric set containerd exposes (will there need to be aliases/will the community support that?)

I also agree that this is a good opportunity for us to cleanup and properly define the metrics endpoint to prepare us for the future

Reply:
I think this is a really good point, this is a great opportunity for us to step back and more clearly define container / pod workload metric endpoints and define more close standardization of metrics across container runtimes.

As a concrete example, containerd already exposes quite a few Prometheus container metrics (see [here](https://gist.github.com/bobbypage/9569bb4a1d3a9d1b3894fd6eab6d3149)) many of which overlap with the existing cadvisor metrics from `/metrics/cadvisor`. 

We can more closely see which metrics are missing from containerd compared to cAdvisor and use that info + community feedback to form a list of standardized metrics that we agree should be exposed by container runtimes in general.

The downside here is of course there will some migration effort required for users. As @Random-Liu commented however, a full migration story involving ability to leave the `/metrics/cadvisor` endpoint on with some feature gate over course of a long deprecation period + offering alternatives of new container runtime metric endpoint or dedicated cAdvisor deamonset seems like a reasonable strategy.

Reply:
> I think it would be ideal to have the label values unchanged from cadvisor, so a user could plug and play the old URL+endpoint with the new one.

Yeah, we can review the current labels, and keep them as long as they make sense.

> Do you see that working with the existing metric set containerd exposes (will there need to be aliases/will the community support that?)

If the metrics defined in the CRI is different from the containerd metrics, I think we can still implement the CRI defined metrics in the containerd CRI plugin. Like I don't think containerd's native metrics endpoint has pod metrics, but it does make sense for CRI.